client/state/data-layer/http/index.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * External dependencies
+ */
+import superagent from 'superagent';
+
+/***
+ * Internal dependencies
+ */
+import { extendAction } from 'state/utils';
+import { HTTP_REQUEST } from 'state/action-types';
+import { failureMeta, successMeta } from 'state/data-layer/wpcom-http';
+
+/***
+ * Handler to perorm an http request based on `HTTP_REQUEST` action parameters:
+ * {String} url the url to request
+ * {String} method the method we should use in the request: GET, POST etc.
+ * {Array<{Object}>} headers array of { key: '', value: '' } pairs for the request headers
+ * {Object} body data send as body
+ * {Boolean} withCredentials save cookie set on request
+ * {Action} onSuccess action to dispatch on success with data meta
+ * {Action} onFailure action to dispatch on failure with error meta
+ *
+ * @param {Function} dispatch redux store dispatch
+ * @param {Function} action dispathed action we need to handle
+ * @returns {Promise} promise of the handled request
+ */
+const httpHandler = ( { dispatch }, action ) => {
+	const {
+		url,
+		method,
+		headers,
+		body,
+		withCredentials,
+		onSuccess,
+		onFailure
+	} = action;
+
+	let request = superagent( method, url );
+
+	if ( withCredentials ) {
+		request = request.withCredentials();
+	}
+
+	if ( Array.isArray( headers ) ) {
+		headers.forEach( header => {
+			if ( header.key && header.value ) {
+				request = request.set( header.key, header.value );
+			}
+		} );
+	}
+
+	request = request.accept( 'application/json' );
```

Comment:
> I have no idea.

so I went ahead and tested this since it wasn't clear in their documentation

```js
const request = require( 'superagent' );
 
let r = request.get( 'https://api.icndb.com/jokes/random')
let s = r.query( 'firstName=John' );
 
r.qsRaw // [ 'firstName=John' ]
s.qsRaw // [ 'firstName=John' ]
```

in fact, all calls in that request API _do_ mutate the original request.

> they can return a new request each time and from the user perspective it would be the same

this is going to be where I disagree because it doesn't return a new request - it returns the old request. I would definitely prefer to let the code indicate this itself so that it's clear what's happening. the reassignment isn't even doing anything here but if someone were to look at it and think they could "save" a partial request and use it else where they would be wrong and that could lead to hard-to-find bugs.

```js
var r1 = request.get( '/path' )
r1 = r1.query( 'something=cool' )

var r2 = r1

r1 = r1.query( 'another=thing' )

// oops! now r2 is ruined because it pointed to the same spot in memory that r1 did and we mutated r1
```